man/gff2xml.1
.Dd September 1, 2016
.Dt GFF2XML 1
.Os
.Sh NAME
.Nm gff2xml
.Nd BioWare GFF to XML converter
.Sh SYNOPSIS
.Nm gff2xml
.Op Ar options
.Ar input_file
.Op Ar output_file
.Sh DESCRIPTION
.Nm
converts BioWare's GFF files (versions V3.2/V3.3 and V4.0/V4.1)
into human-readable XML.
GFF are hierarchical data files, similar
to XML in concept, but stored in binary.
As such, these files are used as a basis for many of the file formats
found in the BioWare games.
For example, an UTC file is a GFF holding a template for a
creature, while a GUI file is a GFF describing an in-game menu.
.Pp
Both version 3 of the format (V3.2/V3.3) and version 4 (V4.0/V4.1)
are supported.
While they are similar, the 4th version carries
several changes to make the files more efficient to read in-game.
This includes replacing the string field names (which map
to XML tags) with numerical identifiers, resulting in converted
XML files that are stripped of their meaning.
To compensate, this tool adds readable aliases to many of these
numerical identifiers, giving them back their meaning.
Unfortunately, not all of them are known.
Most notably, the identifiers introduced in
.Em Sonic Chronicles: The Dark Brotherhood
and
.Em Dragon Age 2
are still missing.
.Pp
The changes in the minor versions (V3.2 vs. V3.3 and V4.0 vs.
V4.1) are less significant.
V3.3 simply changes which languages are supported,
and V4.1 adds a common string table at the start of the file.
Both of these additions are handled transparently.
.Pp
LocStrings found in GFF V3.2 and V3.3 contain localized string
data, which, depending on the game and the language, can be encoded
in various ways.
There is no way to autodetect the specific encoding.
.Nm
employs a simple heuristic to combat this, but it may fail
for certain strings and files.
However, there are options to explicitly specify the game this GFF
file is from.
.Nm
will then use the correct game-specific encoding tables.
.Pp
Unfortunately, even these tables might not be completely correct
in all cases.
.Em Neverwinter Nights ,
for example, treated many strings as being encoded in the native
encoding used for the language of the game installation.
This lead to many people putting non-English strings into fields
tagged as language ID 0, nominally reserved for English.
To read these files correctly,
.Nm
provides an --encoding parameter to override the encoding used
for a specific language ID.
.Sh OPTIONS
.Bl -tag -width xxxx -compact
.It Fl h
.It Fl Fl help
Show a help text and exit.
.It Fl Fl version
Show version information and exit.
.It Fl Fl cp1252
Read GFF4 strings as Windows CP-1252.
Usually, strings in version 4 of the GFF format are encoded in
little-endian UTF-16.
But some files store them as Windows CP-1252 instead.
Since there's no clean way to autodetect the different encoding,
this switch manually selects Windows CP-1252.
This option only concerns strings embedded in GFF4 files, not
GFF3 LocStrings.
.It Fl Fl nwnpremium
The GFF files found in the encrypted HAK files of
.Em Neverwinter Nights
premium modules are deliberately broken.
This options tells
.Nm
to work around the brokenness.
.It Fl Fl nwn
Read LocStrings in an encoding appropriate for
.Em Neverwinter Nights .
.It Fl Fl nwn2
Read LocStrings in an encoding appropriate for
.Em Neverwinter Nights 2 .
.It Fl Fl kotor
Read LocStrings in an encoding appropriate for
.Em Knights of the Old Republic .
.It Fl Fl kotor2
Read LocStrings in an encoding appropriate for
.Em Knights of the Old Republic II .
.It Fl Fl jade
Read LocStrings in an encoding appropriate for
.Em Jade Empire .
.It Fl Fl witcher
Read LocStrings in an encoding appropriate for
.Em The Witcher .
.It Fl Fl dragonage
Read LocStrings in an encoding appropriate for
.Em Dragon Age: Origins .
.It Fl Fl dragonage2
Read LocStrings in an encoding appropriate for
.Em Dragon Age II .
.It Fl Fl encoding Ar str
Override an encoding.
The string has to be of the form n=encoding, for example 0=cp-1252
to override the encoding of the (ungendered) language ID 0 to be
Windows codepage 1252.
To override several encodings, specify the --encoding parameter
multiple times.
.It Fl Fl sac
Assume a header found in SAC files.
.El
.Bl -tag -width xxxx -compact
.It Ar input_file
The GFF file to convert.
.It Op Ar output_file
The XML file will be written there.
If no output file is specified, the XML data is written to
.Dv stdout .
The encoding of the XML stream is always UTF-8.
.El
.Sh EXAMPLES
Convert the GFF
.Pa file1.utc
into an XML file:
.Pp
.Dl $ gff2xml file1.utc file2.xml
.Pp
Convert the GFF
.Pa file1.utc
into an XML file on
.Dv stdout :
.Pp
.Dl $ gff2xml file1.utc
.Pp
Convert the GFF
.Pa file1.utc ,
which uses Windows CP-1252 strings:
.Pp
.Dl $ gff2xml --cp1252 file1.utc file2.xml
.Pp
Convert the GFF
.Pa file1.utc ,
which encodes language ID 0 in LocStrings as Windows CP-1250:
.Dl $ gff2xml --encoding 0=cp1250 file1.utc file2.xml
.Sh SEE ALSO
.Xr xml2gff 1 ,
.Xr convert2da 1 ,
.Xr fixpremiumgff 1 ,
.Xr tlk2xml 1 ,
.Xr ssf2xml 1
.Pp
More information about the xoreos project can be found on
.Lk https://xoreos.org/ "its website"
.Ns .
.Sh AUTHORS
This program is part of the xoreos-tools package, which in turn is
part of the xoreos project, and was written by the xoreos team.
Please see the
.Pa AUTHORS
file for details.


NEWS
Changes in xoreos-tools version 0.0.2
=====================================

This is the first official release of xoreos-tools, together with xoreos.

Included are:

- Tools for handling resource archives
  - unkeybif
  - unerf
  - unrim
  - unherf
  - unnds
  - desmall
- Tools converting basic resource formats into human-readable form
  - gff2xml
  - tlk2xl
  - convert2da
- Tools converting graphics formats into TGA
  - xoreostex2tga
  - unnsbtx
  - nbfs2tga
  - ncgr2tga
  - cbgt2tga
  - cdpth2tga


src/gff2xml.cpp
/* xoreos-tools - Tools to help with xoreos development
 *
 * xoreos-tools is the legal property of its developers, whose names
 * can be found in the AUTHORS file distributed with this source
 * distribution.
 *
 * xoreos-tools is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or (at your option) any later version.
 *
 * xoreos-tools is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with xoreos-tools. If not, see <http://www.gnu.org/licenses/>.
 */

/** @file
 *  Tool to convert GFF files into XML.
 */

#include <cstring>
#include <cstdio>

#include "src/version/version.h"

#include "src/common/scopedptr.h"
#include "src/common/ustring.h"
#include "src/common/util.h"
#include "src/common/strutil.h"
#include "src/common/error.h"
#include "src/common/platform.h"
#include "src/common/readfile.h"
#include "src/common/writefile.h"
#include "src/common/stdoutstream.h"
#include "src/common/encoding.h"
#include "src/common/cli.h"

#include "src/aurora/types.h"
#include "src/aurora/language.h"

#include "src/xml/gffdumper.h"

#include "src/util.h"

typedef std::map<uint32, Common::Encoding> EncodingOverrides;

bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue,
                      Common::UString &inFile, Common::UString &outFile,
                      Common::Encoding &encoding, Aurora::GameID &game,
                      EncodingOverrides &encOverrides, bool &nwnPremium, bool &sacFile);

bool parseEncodingOverride(const Common::UString &arg, EncodingOverrides &encOverrides);

void dumpGFF(const Common::UString &inFile, const Common::UString &outFile, Common::Encoding encoding, bool nwnPremium,
             bool sacFile);

int main(int argc, char **argv) {
	initPlatform();

	try {
		std::vector<Common::UString> args;
		Common::Platform::getParameters(argc, argv, args);

		Common::Encoding encoding = Common::kEncodingInvalid;
		Aurora::GameID   game     = Aurora::kGameIDUnknown;

		EncodingOverrides encOverrides;

		bool nwnPremium = false;
		bool sacFile = false;

		int returnValue = 1;
		Common::UString inFile, outFile;

		if (!parseCommandLine(args, returnValue, inFile, outFile, encoding, game, encOverrides, nwnPremium, sacFile))
			return returnValue;

		LangMan.declareLanguages(game);

		for (EncodingOverrides::const_iterator e = encOverrides.begin(); e != encOverrides.end(); ++e)
			LangMan.overrideEncoding(e->first, e->second);

		dumpGFF(inFile, outFile, encoding, nwnPremium, sacFile);
	} catch (...) {
		Common::exceptionDispatcherError();
	}

	return 0;
}

bool parseEncodingOverride(const Common::UString &arg, EncodingOverrides &encOverrides) {
	Common::UString::iterator sep = arg.findFirst('=');
	if (sep == arg.end())
		return false;

	uint32 id = 0xFFFFFFFF;
	try {
		Common::parseString(arg.substr(arg.begin(), sep), id);
	} catch (...) {
	}

	if (id == 0xFFFFFFFF)
		return false;

	Common::Encoding encoding = Common::parseEncoding(arg.substr(++sep, arg.end()));
	if (encoding == Common::kEncodingInvalid) {
		status("Unknown encoding \"%s\"", arg.substr(sep, arg.end()).c_str());
		return false;
	}

	encOverrides[id] = encoding;
	return true;
}

bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue,
                      Common::UString &inFile, Common::UString &outFile,
                      Common::Encoding &encoding, Aurora::GameID &game,
                      EncodingOverrides &encOverrides, bool &nwnPremium, bool &sacFile) {
	using Common::CLI::NoOption;
	using Common::CLI::kContinueParsing;
	using Common::CLI::Parser;
	using Common::CLI::ValGetter;
	using Common::CLI::Callback;
	using Common::CLI::ValAssigner;
	using Common::CLI::makeEndArgs;
	using Common::CLI::makeAssigners;
	using Aurora::GameID;

	NoOption inFileOpt(false, new ValGetter<Common::UString &>(inFile, "input file"));
	NoOption outFileOpt(true, new ValGetter<Common::UString &>(outFile, "output file"));
	Parser parser(argv[0], "BioWare GFF to XML converter",
	              "If no output file is given, the output is written to stdout.\n\n"
	              "Depending on the game, LocStrings in GFF files might be encoded in various\n"
	              "ways and there's no way to autodetect how. If a game is specified, the\n"
	              "encoding tables for this game are used. Otherwise, gff2xml tries some\n"
	              "heuristics that might fail for certain strings.\n\n"
	              "Additionally, the --encoding parameter can be used to override the encoding\n"
	              "for a specific language ID. The string has to be of the form n=encoding,\n"
	              "for example 0=cp-1252 to override the encoding of the (ungendered) language\n"
	              "ID 0 to be Windows codepage 1252. To override several encodings, specify\n"
	              "the --encoding parameter multiple times.\n",
	              returnValue,
	              makeEndArgs(&inFileOpt, &outFileOpt));


	parser.addSpace();
	parser.addOption("cp1252", "Read GFF4 strings as Windows CP-1252", kContinueParsing,
	                 makeAssigners(new ValAssigner<Common::Encoding>(Common::kEncodingCP1252,
	                 encoding)));
	parser.addSpace();
	parser.addOption("nwnpremium", "This is a broken GFF from a Neverwinter Nights premium module",
	                 kContinueParsing,
	                 makeAssigners(new ValAssigner<bool>(true, nwnPremium)));
	parser.addSpace();
	parser.addOption("nwn", "Use Neverwinter Nights encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDNWN, game)));
	parser.addOption("nwn2", "Use Neverwinter Nights 2 encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDNWN2, game)));
	parser.addOption("kotor", "Use Knights of the Old Republic encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDKotOR, game)));
	parser.addOption("kotor2", "Use Knights of the Old Republic II encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDKotOR2, game)));
	parser.addOption("jade", "Use Jade Empire encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDJade, game)));
	parser.addOption("witcher", "Use The Witcher encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDWitcher, game)));
	parser.addOption("dragonage", "Use Dragon Age encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDDragonAge, game)));
	parser.addOption("dragonage2", "Use Dragon Age II encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDDragonAge2, game)));
	parser.addSpace();
	parser.addOption("encoding", "Override an encoding", kContinueParsing,
	                 new Callback<EncodingOverrides &>("str", parseEncodingOverride, encOverrides));
	parser.addOption("sac", "Read the extra sac file header", kContinueParsing,
	                 makeAssigners(new ValAssigner<bool>(true, sacFile)));

	return parser.process(argv);
}


void dumpGFF(const Common::UString &inFile, const Common::UString &outFile, Common::Encoding encoding, bool nwnPremium,
             bool sacFile) {

	Common::ScopedPtr<Common::SeekableReadStream> gff(new Common::ReadFile(inFile));

	Common::ScopedPtr<XML::GFFDumper> dumper(XML::GFFDumper::identify(*gff, nwnPremium, sacFile));

	Common::ScopedPtr<Common::WriteStream> out(openFileOrStdOut(outFile));

	dumper->dump(*out, gff.release(), encoding, nwnPremium);

	out->flush();

	if (!outFile.empty())
		status("Converted \"%s\" to \"%s\"", inFile.c_str(), outFile.c_str());
}
